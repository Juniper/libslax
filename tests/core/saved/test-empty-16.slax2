version 1.2;

param $full;
var $bad;
var $good = "first";
var $data = <data> "\nthis is line one\nand line two\n\nand line three\nand line four\nand finally, line five\n";

main <top> {
    <back-reference> {
        /*
         * The "BUGS" section of the FreeBSD man page for regexec(3) says:
         *     The back-reference code is subtle and doubts linger
         *     about its  correctness in complex cases.
         * I guess this code echoes that since it fails under FB.
         */
        var $pat = "([a-z]):\\1:\\1";
        var $re1 = slax:regex($pat, "a:a:a");
        
        <re1> {
            copy-of $re1;
        }
        var $re2 = slax:regex($pat, "a:b:c");
        <re2> {
            copy-of $re2;
        }
        var $re3 = slax:regex($pat, "a:b:b:b:a:a");
        <re3> {
            copy-of $re3;
        }
    }
    <sequence> {
        for-each (1 ... 3) {
            for $i (1 ... 4) {
                <i> . _ ":" _ $i;
            }
            
        }
    }
    <first> slax:first-of($bad, worse, $good, "huh");
    <break-lines> {
        var $bl = slax:break-lines($data);
        
        for-each ($bl) {
            <line> .;
        }
    }
    <printf> {
        for-each (1 ... 4) {
            var $boom = {
                if (position() == last()) {
                    expr " boom!!";
                }
            }
            
            <one> slax:printf("testing %j1-4d... %4s... %s%s!", 1, . + 1, "three", $boom);
        }
        <two> slax:printf("[%jcs] [%jt{test}s] [%jt{test}s]", "cool", "ing", no);
    }
    <is-empty> {
        var $e1;
        var $e2 := <test>;
        var $ne1 = "test";
        var $ne2 = <test>;
        mvar $ne3;
        
        <e1> slax:is-empty($e1);
        <e2> slax:is-empty($e2/fish);
        <e1> slax:is-empty($ne3);
        <ne1> slax:is-empty($ne1);
        <ne2> slax:is-empty($ne2);
        set $ne3 = <test>;
        <ne3> slax:is-empty($ne3);
    }
    <regex> {
        var $data = "this is a test of the regex";
        var $pattern = "(.*) test (.*) regex";
        var $re1 = slax:regex($pattern, $data);
        
        for-each ($re1) {
            <re1 position=position()> .;
        }
        var $re2 = slax:regex("([0-9]+)", $data);
        
        for-each ($re2) {
            <re2 position=position()> .;
        }
        var $re3 = slax:regex("([0-9]+)", "So..... 99 bottles of ... ");
        
        for-each ($re3) {
            <re3 position=position()> .;
        }
    }
    <split> {
        var $in = "instructions 1 wash 2 rinse 3 repeat 4 not-reached";
        var $sp1 = slax:split("[0-9]", $in);
        
        for-each ($sp1) {
            <sp1 position=position()> .;
        }
        var $sp2 = slax:split("[0-9]", $in, 4);
        
        for-each ($sp2) {
            <sp2 position=position()> .;
        }
    }
    if ($full) {
        <dampen> {
            for-each (1 ... 100) {
                if (slax:dampen("test", 4, 5)) {
                    <works> .;
                }
            }
        }
        <sleep> {
            message "before";
            expr slax:sleep(3);
            message "after";
        }
        <sysctl> {
            var $names := {
                <v> "kern.ostype";
                <v> "kern.osrelease";
                <v> "kern.hostname";
            }
            
            for-each ($names/v) {
                <var name=.> slax:sysctl(.);
            }
            <var name="kern.maxproc"> slax:sysctl("kern.maxproc", "i");
        }
        <syslog> slax:syslog("daemon.error", "test", " message");
    }
}
