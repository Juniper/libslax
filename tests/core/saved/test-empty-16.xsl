<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform" xmlns:slax="http://xml.libslax.org/slax" xmlns:slax-ext="http://xmlsoft.org/XSLT/namespace" version="1.0" extension-element-prefixes="slax slax-ext">
  <xsl:param name="full"/>
  <xsl:variable name="bad"/>
  <xsl:variable name="good" select="&quot;first&quot;"/>
  <xsl:variable name="data">
    <data>
this is line one
and line two

and line three
and line four
and finally, line five
</data>
  </xsl:variable>
  <!-- 
 * Turns out, back references aren't part of the posix standard, so
 * including them in tests is a Bad Idea (tm).
 -->
  <xsl:param name="force" select="0"/>
  <xsl:template match="/">
    <top>
      <xsl:if test="$force">
        <back-reference>
          <!-- 
	         * The "BUGS" section of the FreeBSD man page for regexec(3) says:
	         *     The back-reference code is subtle and doubts linger
	         *     about its  correctness in complex cases.
                 * Not sure what this means, but this fails under MacOS, while
                 * working correctly under FreeBSD and Linux.  Perl likes it also.
 -->
          <xsl:variable name="pat" select="&quot;([a-z]):\1:\1&quot;"/>
          <xsl:variable xmlns:slax="http://xml.libslax.org/slax" name="re1" select="slax:regex($pat, &quot;a:a:a&quot;)"/>
          <re1>
            <xsl:copy-of select="$re1"/>
          </re1>
          <xsl:variable xmlns:slax="http://xml.libslax.org/slax" name="re2" select="slax:regex($pat, &quot;a:b:c&quot;)"/>
          <re2>
            <xsl:copy-of select="$re2"/>
          </re2>
          <xsl:variable xmlns:slax="http://xml.libslax.org/slax" name="re3" select="slax:regex($pat, &quot;a:b:b:b:a:a&quot;)"/>
          <re3>
            <xsl:copy-of select="$re3"/>
          </re3>
        </back-reference>
      </xsl:if>
      <multiple>
        <xsl:variable name="pat" select="&quot;([a-z]):&quot;"/>
        <xsl:variable xmlns:slax="http://xml.libslax.org/slax" name="re1" select="slax:regex($pat, &quot;a:a:a&quot;)"/>
        <re1>
          <xsl:copy-of select="$re1"/>
        </re1>
        <xsl:variable xmlns:slax="http://xml.libslax.org/slax" name="re2" select="slax:regex($pat, &quot;a:b:c&quot;)"/>
        <re2>
          <xsl:copy-of select="$re2"/>
        </re2>
        <xsl:variable xmlns:slax="http://xml.libslax.org/slax" name="re3" select="slax:regex($pat, &quot;a:b:b:b:a:a&quot;)"/>
        <re3>
          <xsl:copy-of select="$re3"/>
        </re3>
      </multiple>
      <sequence>
        <xsl:for-each xmlns:slax="http://xml.libslax.org/slax" select="slax:build-sequence(1, 3)">
          <xsl:variable name="slax-dot-1" select="."/>
          <xsl:for-each xmlns:slax="http://xml.libslax.org/slax" select="slax:build-sequence(1, 4)">
            <xsl:variable name="i" select="."/>
            <xsl:for-each select="$slax-dot-1">
              <i>
                <xsl:value-of select="concat(., &quot;:&quot;, $i)"/>
              </i>
            </xsl:for-each>
          </xsl:for-each>
        </xsl:for-each>
      </sequence>
      <first>
        <xsl:value-of xmlns:slax="http://xml.libslax.org/slax" select="slax:first-of($bad, worse, $good, &quot;huh&quot;)"/>
      </first>
      <break-lines>
        <xsl:variable xmlns:slax="http://xml.libslax.org/slax" name="bl" select="slax:break-lines($data)"/>
        <xsl:for-each select="$bl">
          <line>
            <xsl:value-of select="."/>
          </line>
        </xsl:for-each>
      </break-lines>
      <printf>
        <xsl:for-each xmlns:slax="http://xml.libslax.org/slax" select="slax:build-sequence(1, 4)">
          <xsl:variable name="boom">
            <xsl:if test="position() = last()">
              <xsl:text> boom!!</xsl:text>
            </xsl:if>
          </xsl:variable>
          <one>
            <xsl:value-of xmlns:slax="http://xml.libslax.org/slax" select="slax:printf(&quot;testing %j1-4d... %4s... %s%s!&quot;, 1, . + 1, &quot;three&quot;, $boom)"/>
          </one>
        </xsl:for-each>
        <two>
          <xsl:value-of xmlns:slax="http://xml.libslax.org/slax" select="slax:printf(&quot;[%jcs] [%jt{test}s] [%jt{test}s]&quot;, &quot;cool&quot;, &quot;ing&quot;, no)"/>
        </two>
      </printf>
      <is-empty>
        <xsl:variable name="e1"/>
        <xsl:variable name="e2-temp-1">
          <test/>
        </xsl:variable>
        <xsl:variable xmlns:slax-ext="http://xmlsoft.org/XSLT/namespace" name="e2" select="slax-ext:node-set($e2-temp-1)"/>
        <xsl:variable name="ne1" select="&quot;test&quot;"/>
        <xsl:variable name="ne2">
          <test/>
        </xsl:variable>
        <xsl:variable name="slax-ne3" mvarname="ne3"/>
        <xsl:variable xmlns:slax="http://xml.libslax.org/slax" name="ne3" mutable="yes" select="slax:mvar-init(&quot;ne3&quot;, &quot;slax-ne3&quot;, $slax-ne3)" svarname="slax-ne3"/>
        <e1>
          <xsl:value-of xmlns:slax="http://xml.libslax.org/slax" select="slax:is-empty($e1)"/>
        </e1>
        <e2>
          <xsl:value-of xmlns:slax="http://xml.libslax.org/slax" select="slax:is-empty($e2/fish)"/>
        </e2>
        <e1>
          <xsl:value-of xmlns:slax="http://xml.libslax.org/slax" select="slax:is-empty($ne3)"/>
        </e1>
        <ne1>
          <xsl:value-of xmlns:slax="http://xml.libslax.org/slax" select="slax:is-empty($ne1)"/>
        </ne1>
        <ne2>
          <xsl:value-of xmlns:slax="http://xml.libslax.org/slax" select="slax:is-empty($ne2)"/>
        </ne2>
        <slax:set-variable xmlns:slax="http://xml.libslax.org/slax" name="ne3" svarname="slax-ne3">
          <test/>
        </slax:set-variable>
        <ne3>
          <xsl:value-of xmlns:slax="http://xml.libslax.org/slax" select="slax:is-empty($ne3)"/>
        </ne3>
      </is-empty>
      <regex>
        <xsl:variable name="data" select="&quot;this is a test of the regex&quot;"/>
        <xsl:variable name="pattern" select="&quot;(.*) test (.*) regex&quot;"/>
        <xsl:variable xmlns:slax="http://xml.libslax.org/slax" name="re1" select="slax:regex($pattern, $data)"/>
        <xsl:for-each select="$re1">
          <re1 position="{position()}">
            <xsl:value-of select="."/>
          </re1>
        </xsl:for-each>
        <xsl:variable xmlns:slax="http://xml.libslax.org/slax" name="re2" select="slax:regex(&quot;([0-9]+)&quot;, $data)"/>
        <xsl:for-each select="$re2">
          <re2 position="{position()}">
            <xsl:value-of select="."/>
          </re2>
        </xsl:for-each>
        <xsl:variable xmlns:slax="http://xml.libslax.org/slax" name="re3" select="slax:regex(&quot;([0-9]+)&quot;, &quot;So..... 99 bottles of ... &quot;)"/>
        <xsl:for-each select="$re3">
          <re3 position="{position()}">
            <xsl:value-of select="."/>
          </re3>
        </xsl:for-each>
      </regex>
      <split>
        <xsl:variable name="in" select="&quot;instructions 1 wash 2 rinse 3 repeat 4 not-reached&quot;"/>
        <xsl:variable xmlns:slax="http://xml.libslax.org/slax" name="sp1" select="slax:split(&quot;[0-9]&quot;, $in)"/>
        <xsl:for-each select="$sp1">
          <sp1 position="{position()}">
            <xsl:value-of select="."/>
          </sp1>
        </xsl:for-each>
        <xsl:variable xmlns:slax="http://xml.libslax.org/slax" name="sp2" select="slax:split(&quot;[0-9]&quot;, $in, 4)"/>
        <xsl:for-each select="$sp2">
          <sp2 position="{position()}">
            <xsl:value-of select="."/>
          </sp2>
        </xsl:for-each>
      </split>
      <xsl:if test="$full">
        <dampen>
          <xsl:for-each xmlns:slax="http://xml.libslax.org/slax" select="slax:build-sequence(1, 100)">
            <xsl:if xmlns:slax="http://xml.libslax.org/slax" test="slax:dampen(&quot;test&quot;, 4, 5)">
              <works>
                <xsl:value-of select="."/>
              </works>
            </xsl:if>
          </xsl:for-each>
        </dampen>
        <sleep>
          <xsl:message>before</xsl:message>
          <xsl:value-of xmlns:slax="http://xml.libslax.org/slax" select="slax:sleep(3)"/>
          <xsl:message>after</xsl:message>
        </sleep>
        <sysctl>
          <xsl:variable name="names-temp-2">
            <v>kern.ostype</v>
            <v>kern.osrelease</v>
            <v>kern.hostname</v>
          </xsl:variable>
          <xsl:variable xmlns:slax-ext="http://xmlsoft.org/XSLT/namespace" name="names" select="slax-ext:node-set($names-temp-2)"/>
          <xsl:for-each select="$names/v">
            <var name="{.}">
              <xsl:value-of xmlns:slax="http://xml.libslax.org/slax" select="slax:sysctl(.)"/>
            </var>
          </xsl:for-each>
          <var name="kern.maxproc">
            <xsl:value-of xmlns:slax="http://xml.libslax.org/slax" select="slax:sysctl(&quot;kern.maxproc&quot;, &quot;i&quot;)"/>
          </var>
        </sysctl>
        <syslog>
          <xsl:value-of xmlns:slax="http://xml.libslax.org/slax" select="slax:syslog(&quot;daemon.error&quot;, &quot;test&quot;, &quot; message&quot;)"/>
        </syslog>
      </xsl:if>
    </top>
  </xsl:template>
</xsl:stylesheet>
