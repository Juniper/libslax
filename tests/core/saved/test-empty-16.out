<?xml version="1.0"?>
<top>
  <multiple>
    <re1>
      <match>a:</match>
      <match>a</match>
    </re1>
    <re2>
      <match>a:</match>
      <match>a</match>
    </re2>
    <re3>
      <match>a:</match>
      <match>a</match>
    </re3>
  </multiple>
  <sequence>
    <i>1:1</i>
    <i>1:2</i>
    <i>1:3</i>
    <i>1:4</i>
    <i>2:1</i>
    <i>2:2</i>
    <i>2:3</i>
    <i>2:4</i>
    <i>3:1</i>
    <i>3:2</i>
    <i>3:3</i>
    <i>3:4</i>
  </sequence>
  <first>first</first>
  <break-lines>
    <line/>
    <line>this is line one</line>
    <line>and line two</line>
    <line/>
    <line>and line three</line>
    <line>and line four</line>
    <line>and finally, line five</line>
  </break-lines>
  <printf>
    <one>testing 1   ...    2... three!</one>
    <one>testing     ...    3... three!</one>
    <one>testing     ...    4... three!</one>
    <one>testing     ...    5... three boom!!!</one>
    <two>[Cool] [testing] []</two>
  </printf>
  <is-empty>
    <e1>true</e1>
    <e2>true</e2>
    <e1>true</e1>
    <ne1>false</ne1>
    <ne2>false</ne2>
    <ne3>false</ne3>
  </is-empty>
  <regex>
    <re1 position="1">this is a test of the regex</re1>
    <re1 position="2">this is a</re1>
    <re1 position="3">of the</re1>
    <re3 position="1">99</re3>
    <re3 position="2">99</re3>
  </regex>
  <split>
    <sp1 position="1">instructions </sp1>
    <sp1 position="2"> wash </sp1>
    <sp1 position="3"> rinse </sp1>
    <sp1 position="4"> repeat </sp1>
    <sp1 position="5"> not-reached</sp1>
    <sp2 position="1">instructions </sp2>
    <sp2 position="2"> wash </sp2>
    <sp2 position="3"> rinse </sp2>
    <sp2 position="4"> repeat 4 not-reached</sp2>
  </split>
</top>
